### PR TITLE
bpo-31245: Asyncio unix socket datagram

### PR DESCRIPTION
**Pull request by ovv**
https://bugs.python.org/issue31245

### CODE REVIEW

File: Lib/asyncio/unix_events.py
Comment (vstinner) [inline Lib/asyncio/unix_events.py]: Please chain the exception.

File: Lib/test/test_asyncio/test_selector_events.py
Comment (vstinner) [inline Lib/test/test_asyncio/test_selector_events.py]: Please use self.sock_fd

File: Lib/asyncio/selector_events.py
Comment (vstinner) [inline Lib/asyncio/selector_events.py]: Are you sure about the semantics? Can't you use send ("") as a cheap ping for example?

I suggest to remove this test.

File: Lib/test/test_asyncio/test_selector_events.py
Comment (vstinner) [inline Lib/test/test_asyncio/test_selector_events.py]: Again, use self.sock_fd

File: Lib/test/test_asyncio/test_unix_events.py
Comment (vstinner) [inline Lib/test/test_asyncio/test_unix_events.py]: I would prefer to check the error code rather than the message.

File: Misc/NEWS.d/next/Library/2017-08-22-11-05-35.bpo-31245.AniZuz.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2017-08-22-11-05-35.bpo-31245.AniZuz.rst]: Please mention asyncio. Maybe "asyncio: ...".

File: Lib/asyncio/unix_events.py
Comment (ovv) [reply Lib/asyncio/unix_events.py]: This was made to improve the error message by including the address that failed the bind. I'm not sure chaining the exception will provide more information on the error.

File: Lib/test/test_asyncio/test_unix_events.py
Comment (ovv) [reply Lib/test/test_asyncio/test_unix_events.py]: I will check both. So we make sure the error message is also correctly rewritten

File: Lib/asyncio/selector_events.py
Comment (ovv) [reply Lib/asyncio/selector_events.py]: It seems to be working without the check. I wonder if it's also not necessary for the `SelectorDatagramTransport`

Comment (njsmith) [reply Lib/asyncio/selector_events.py]: That's correct – on a datagram socket, sending a packet with empty payload is totally legal, and is different from sending no packet at all.

**Comment by njsmith** [comment]
This is a *ton* of new code (mostly copy/pasted?) for what's basically just... changing one constant. At the very least can't you share `DatagramProtocol` and most of the test code?

**Comment by 1st1** [comment]
> This is a ton of new code (mostly copy/pasted?) for what's basically just... changing one constant. At the very least can't you share DatagramProtocol and most of the test code?

+1. This should be refactored, and I wouldn't expect to see a patch that's longer than +- 50 lines of code.  Also, please don't merge without my final review.

**Comment by ovv** [comment]
Thanks for the feedback. I'll try to refactor to shorten the patch.

Concerning the protocol I'm not sure on the direction to take. I could use the `DatagramProtocol`, but in my opinion it doesn't make much sense to pass the `addr` to `datagram_received` for a unix socket.

**Comment by ovv** [comment]
Well it appears I had some misconception about how unix datagram socket works. 

I think this patch is much more reasonable. Sorry for the mess it was previously.

File: Lib/asyncio/base_events.py
Comment (1st1) [inline Lib/asyncio/base_events.py]: Make it a proper exception -- `TypeError` in this case.

**Review by 1st1 — CHANGES_REQUESTED** [review]
Please resolve one remaining nit.  Otherwise LGTM

**Comment by ovv** [comment]
I have made the requested changes; please review again

**Review by asvetlov — APPROVED** [review]

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
Thanks!

**Pull request merged**
